Sources/HummingbirdWebSocket/WebSocketOutboundWriter.swift

```diff
@@ -50,4 +50,59 @@ public struct WebSocketOutboundWriter: Sendable {
             try await self.handler.write(frame: frame)
         }
     }
+
+    struct MessageWriter {
+        let handler: WebSocketHandler
+        var prevFrame: WebSocketFrame?
+        mutating func write(_ data: ByteBuffer, opcode: WebSocketOpcode) async throws {
+            if let prevFrame {
+                try await self.handler.write(frame: prevFrame)
+                self.prevFrame = .init(fin: false, opcode: .continuation, data: data)
+            } else {
+                self.prevFrame = .init(fin: false, opcode: opcode, data: data)
+            }
+        }
+
+        func finish() async throws {
+            if var prevFrame {
+                prevFrame.fin = true
+                try await self.handler.write(frame: prevFrame)
+            }
+        }
+    }
+
+    /// Write a single WebSocket text message as a series of fragmented frames
+    /// - Parameter write: Function writing frames
+    public func withTextMessageWriter<Value>(_ write: ((String) async throws -> Void) async throws -> Value) async throws -> Value {
+        var writer = MessageWriter(handler: self.handler)
+        let value: Value
+        do {
+            value = try await write { text in
+                let data = writer.handler.context.allocator.buffer(string: text)
+                try await writer.write(data, opcode: .text)
+            }
+        } catch {
+            try await writer.finish()
+            throw error
+        }
+        try await writer.finish()
+        return value
+    }
+
+    /// Write a single WebSocket binary message as a series of fragmented frames
+    /// - Parameter write: Function writing frames
+    public func withBinaryMessageWriter<Value>(_ write: ((ByteBuffer) async throws -> Void) async throws -> Value) async throws -> Value {
```

Comment:
This could leverage the same writer mentioned above, but the writer would store the expected opcode. I think it's valid to send a `String` over a binary WebSocket message.

Reply:
done